x-pack/plugin/ml/log-structure-finder/src/main/java/org/elasticsearch/xpack/ml/logstructure/AbstractStructuredLogFileStructureFinder.java

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.ml.logstructure;
+
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.xpack.ml.logstructure.TimestampFormatFinder.TimestampMatch;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+import java.util.SortedMap;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+public abstract class AbstractStructuredLogFileStructureFinder extends AbstractLogFileStructureFinder {
+
+    protected Tuple<String, TimestampMatch> guessTimestampField(List<String> explanation, List<Map<String, ?>> sampleRecords) {
```

Comment:
While adding the comment I noticed that for ordered structured formats with multiple timestamps we weren't consistently taking the first timestamp in left-to-right order in the original data (due to using a `HashMap` rather than a `LinkedHashMap`).  So I fixed that too.